File: src/RemoveDuplicates.java

//Input: nums = [0,0,1,1,1,1,2,3,3]
//Output: 7, nums = [0,0,1,1,2,3,3,_,_]
//TC O(n)
//SC O(1)
public class RemoveDuplicates {
    public int removeDuplicates(int[] nums) {
        int i = 0;
        for (int num :nums) {
            if (i<2||num>nums[i-2]){
                nums[i++] = num;
            }
        }
        return i;
    }
}

File: RemoveDuplicates.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ConsoleApp1
{
    public class RemoveDuplicates
    {
        /*
         *  T.C: O(n) since traversing through whole array 
         *  S.C: O(1) not using extra space
         */

        public int RemoveDuplicatesFromArray(int[] nums)
        {

            if (nums == null || nums.Length == 0)
                return 0;

            int slow = 1;

            int counter = 1;

            for (int fast = 1; fast < nums.Length; fast++)
            {
                if (nums[fast] == nums[fast - 1])
                {
                    counter++;
                }
                else
                {
                    counter = 1;
                }

                if (counter <= 2)
                {
                    nums[slow] = nums[fast];
                    slow++;
                }
            }

            return slow;
        }
    }
}
